### ISSUE TITLE
[RFC] Functional API design for `torch.nn.Module`

### ISSUE THREAD

**Issue by emcastillo** [pytorch/pytorch#58839]
## 🚀 Feature

Add support for `torch.nn.Module` to be used in functional style.
This RFC proposes a design for https://github.com/pytorch/pytorch/issues/49171

## Motivation & Pitch

The following points were originally written by @albanD 

The goal would be to be able to use nn.Module in two new context that are currently very tricky:
the high level autograd API. Currently, this kind of hack is needed: https://github.com/pytorch/pytorch/blob/2832e325ddad1676ef7aae75da5bfd8e6713e498/benchmarks/functional_autograd_benchmark/utils.py#L19-L71

To be able to make libraries like higher (https://github.com/facebookresearch/higher) much simpler to implement and easy to hack into.

We don't really have a design for it right now but the main idea would be to provide an API to convert between nn.Module and a pair of (function, params_tuple). Where the function would take as input both the nn.Module's regular input and the params_tuple.

make sure it works well with functional autograd, DP and vmap and maybe DDP as well if they want to rework it.
The new vmap+autograd prototype from https://github.com/zou3519/functorch also needs that

### Currently being used hacks

Detailed in https://github.com/pytorch/pytorch/blob/2832e325ddad1676ef7aae75da5bfd8e6713e498/benchmarks/functional_autograd_benchmark/utils.py#L19-L71

- Extracts the module parameters with their names. (The module will be empty), but the parameters are converted to regular tensors
- Creates a new forward function that loads these parameters back and just runs the regular forward pass, but they are loaded as tensors for the forward to work, not as parameters. 

Example pseudocode of how to utilize this:

```python
model = MyCoolTorchModule()  # This is a module that holds parameters

def get_functional_module(module):
    tensors, names = _recursive_remover(module)
    # Parameters have been deleted
    assert len(model.parameters()) == 0
    def functional_fwd(*new_params):
       # helper function that sets the parameters back
       _recursive_adder(module, names, new_params)
       return module.forward(inputs)
    return tensors, function

parameters, model_fn = get_functional_module(model)

# Functional use of the above
grad_fn = torch.functional_api.grad(model_fn)
gradn_fn(parameters, inputs)
```

The current approach shows the following drawbacks

- Needs to remove the module parameters
- Does not restore them, once the module gets its parameters removed then it won’t be able to be used as a regular module again. state_dict and other functions relying on module parameters won’t work. 
- Needs to maintain state for a single model in different places
- External objects and functions will access `torch.nn.Module` internals. Not clean

## Proposal

Add a `torch.nn.Module.functional()` method similar to `.eval()` and `.train()`.
- The module will manage its own status without having an external class/function
Accessing its internals
- `torch.nn.Module` becomes big & difficult to maintain.
- Allows to convert back/from functional state a single model
- Adding another mode to `torch.nn.Module` requires to deal with 3 possible configurations and switches between them. eval/train/functional, this can lead to issues and be a potential no-compat for user-defined modules
- Overhead of setting the parameters back and from every iteration can be high.
- If the  parameters that are passed functionally to the module are not views of the original parameters, in-place updates to these will not have any effect on the current module state

```python
# torch/nn/module.py

class _FunctionalState:
    def __init__(self, parameters):
        self._names, self._parameters = list(parameters)
    def __iter__(self):
        return self
    def __next__(self):
        for n, p in zip(self._names, self._parameters):
            yield n, p

class Module:

    def functional(functional=True):
       # helper function that sets a hook to replace the parameters
       # pre and after the forward
       func_state = None
       def pre_hook(self, module, inputs):
           nonlocal func_state
           func_state = _FunctionalState(self.named_parameters())    
           # assume that parameters are the first n entries of inputs
           for name, param in zip(func_state._names, inputs):
               module._replace_param(name, param)

       def after_hook(self, module, inputs, outputs):
           for name, param in func_state:
               module._replace_param(name, param)
       if functional:
           self._func_pre_hook = self.register_forward_pre_hook(pre_hook)
           self._func_hook = self.register_forward_hook(after_hook)
       else:
           self._func_pre_hook.remove()
           self._func_hook.remove()
       # return parameters as tensors
       return tuple(p.detach().requires_grad_() for p in self.parameters())


# Functional use of the above
model = MyCoolTorchModule()
parameters = model.functional()
jvp_fn = torch.functional_api.jvp(model)
jvp_fn(parameters, inputs)
```

If extending module is not acceptable, we can implement a similar solution in a wrapper class
```python
 class FunctionalModule(torch.nn.Module):
     def __init__(self, module):
           self.module = module
           self.func_state = None
           def pre_hook(self, module, inputs):
               module.func_state = _FunctionalState(self.named_parameters())    
               # assume that parameters are the first n entries of inputs
               for name, param in zip(module.func_state._names, inputs):
                   module._replace_param(name, param)
           def after_hook(self, module, inputs, outputs):
               for name, param in module.func_state:
                   module._replace_param(name, param)
           self._func_pre_hook = self.register_forward_pre_hook(pre_hook)
           self._func_hook = self.register_forward_hook(after_hook)

     def parameters(self):
          return tuple(p.detach.requires_grad_() for p in self.module.parameters())

# Functional use of the above
model = torch.nn.FunctionalModule(MyCoolTorchModule()
parameters = module.parameters()
jvp_fn = torch.functional_api.jvp(model)
jvp_fn(parameters, inputs)
```

Any thoughts? would like some feedback and discuss this in detail :)
cc. @albanD @mruberry @zou3519 @kmaehashi @asi1024 

cc @albanD @jbschlosser @ngimel @mruberry @zou3519

**Comment by zou3519** [comment]
Thanks for the writeup, @emcastillo! I have some quick high-level feedback:

> Currently being used hacks

If you're interested in looking at other previous work,
- I took the functional autograd benchmark code and cleaned it up to provide a `make_functional` and `make_functional_with_buffers` [API](https://github.com/zou3519/functorch/blob/ae97def8eb8508418053a1a7c81371b9b44dcc3d/functorch/_src/make_functional.py#L96-L169)
- [higher also has a variant](https://github.com/facebookresearch/higher/blob/1e20cf9696054277b2d760f64835d5d74a3115a2/higher/patch.py#L195) but I don't know the differences.

> Any thoughts? 

One requirement is that the API should also handle buffers in addition to parameters: one should be able to "rip out" the buffers from a module.

**Comment by emcastillo** [comment]
@zou3519 Thanks a lot for the prompt reply!!!

> Do you have ideas on how optimizers should work with this API?

As far as I can tell right now, the optimizers should work normally on the model-defined parameters since these are being automatically restored after the forward pass is done. But, if we use the optimizer in the tensors that we are using when calling the model in the functional mode, then the optimizer will correctly update the model "real" parameters if the tensors that the functional API uses are views of the parameters themselves. However, I see that there are several use cases where maybe the parameters that we call the functional model with are different tensors from the model real parameters and we want to optimize the former, so a consensus should be reach here.

 > To clarify, does model.functional() change the forward pass of a module? That's what it looks like in your example.

Yes, it adds two hooks, one to set the forward *args tensors as the model parameters before the real forward, and another one to restore the module original parameters after the forward call (should this be needed?)

**Comment by albanD** [comment]
Thanks for writing this down! It looks very interesting.

I think I do agree with you on a few points above:
- If we can avoid baking this into the nn.Module class, it would be better.
- Adding a new builtin state can be confusing (a lot of users already have troubles with train/eval). Also you would want a functional train and a functional eval.
- It would be great if we could re-use existing constructs. In particular, I feel like the things like the [replicate()](https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/replicate.py#L78) function we use for DP has a very similar semantic to what you're doing here (note that this function is quite horrible right now, so I'm not saying we should re-use it directly).
- For the optimizers to "just work" we will need to make sure we never change the Tensor object containing the parameters (whether they are on the Module or in another struct while being functional)

Also I would be curious how the version from higher works? And what are the benefits of theirs (which obviously works in practice).

**Comment by albanD** [comment]
Following some offline discussion with Richard, I think there is another important idea that we want to take into account: we should never create "invalid" nn.Modules.

In particular, the design based on hooks might not verify this because how the Module looks and how it actually behaves during the forward are two different things.
This is a problem we encounter today with the implementation of weight norm for example that makes it impossible to compose with other complex structures like DP.

A possible approach to reduce this problem might be to enforce that we never change the given Module inplace and we simply return an independent function (via a replicate function that copy some objects and leave others empty).
Meaning that you can get multiple functions and they won't interact with each other or the original Module (for example batchnorm state are not shared by default).

**Comment by zou3519** [comment]
For validating the usability of the design, we have a number of use cases and example code over at functorch that you could try to replace with a design and get a sense for how usable and readable it is:
- Per-sample-gradient computation: https://github.com/zou3519/functorch/blob/main/examples/dp_cifar10/cifar10_transforms.py#L88
- Ensemble training: https://github.com/zou3519/functorch/blob/60fbac375187f96d2a652c9b9dcf1f7ef86af360/examples/ensembling/parallel_train.py#L53
- MAML: https://github.com/zou3519/functorch/blob/60fbac375187f96d2a652c9b9dcf1f7ef86af360/examples/maml_regression/evjang_transforms_module.py#L34

**Comment by zou3519** [comment]
Here are some other fun considerations that I ran into recently with functorch:
- There should be a nice way to take a "functional version of a module" and put it into eval mode or train mode. 
- The functional version should work with LazyModules and modules that are lazy (like BatchNorm). In particular: Lazy Modules create state on the first forward pass to the module; we should think about how this interacts with e.g. a user creating a functional version of a module before invoking the forward pass.

**Comment by emcastillo** [comment]
Thank you all for your Thank you all for your feedback,

After thinking about this, I think we can summarize the design goals in the following ones:

1. `torch.nn.Module` must remain untouched.

2. Module must be always in a valid state, this means, while we are operating in functional mode, the original module object must be consistent should its non functional behavior be invoked.

3. This must work with Optimizers & DP.

4. This must work with lazy modules.

5. The functional mode should support both eval/train mode of a module.



1 and 2 introduces the need of a module replicating mechanism that allow to get the original module to remain untouched. Also all the functional behavior should be injected by the function that replicates the original module by using a wrapping/mixin class which contains all the logic for functional management. 
maybe we can have a `torch.nn.FunctionalModule` and make the replicated module inherit from it instead of the regular `torch.nn.Module`

For this approach to work with Optimizers, the replicated module and obtained parameters must be references of the parameters registered in the optimizer (usually the original non-functional module one?), otherwise any inplace operation to these will not update the original module parameters, rendering the optimization useless.

Lazy modules support becomes tricky. Currently the module changes its own parameters during the forward pass to materialize them. If the list of parameters is obtained and `UninitilializedParameters` or buffers are in this list, up to which extent is it ok to alter the input parameter object? I believe that we should require a module to do a dry-run before converting it to a functional module.

I think that replicating the module also solves the eval/train switching, however with functional programming we would like to have a function to work always in the same mode, so maybe we can restrict replication to replicate the module and allow only train or eval. This should be more semantically correct.

Finally, I belive that changes to torch.nn.Module (but unrelated to this) might be needed in order to do a clean implementation and just don't do another monkey patching approach. For example, instead of storing parameters as class members, we can make them live in a `ParameterStorage` object associated to the module, but still access to them through `self.param` by inspecting this object in the `__get/setattr__` methods.

**Comment by zou3519** [comment]
> Higher seems to be doing something like this, a complete module replication with the ability to sync weights.

Could you elaborate a bit more on this? "The ability to sync weights" isn't in the requirements (and I wonder what the use cases for this are)

**Comment by emcastillo** [comment]
> Could you elaborate a bit more on this? "The ability to sync weights" isn't in the requirements you listed (and I wonder what the use cases for this are)

There seems to be this function [higher.patch.buffer_sync](https://higher.readthedocs.io/en/latest/patch.html#higher.patch.buffer_sync) in higher docs
but looking at the code it seems mostly for internal use. I am not sure if we need explicit support for something like this. since a functional module won't hold parameters in our expected design, right?

**Comment by albanD** [comment]
Thanks for all the updates.

On the replicating side, I do agree that it should help quite a lot with having a simpler semantic.
I think we can also make it configurable with flags specifying if the weight/buffers/attributes are shared with the original Module or not.
But I guess that by default, we want the replica to be a standalone version that will behave as if you ran a forward on the Module at that point of time. And so, unless you explicitly ask for sharing, modifications on the module (or on the functional version of self) should not impact the other side.

On that same spirit, I think there are a few core nn features that we want to ensure work fine:
- Lazy modules (forcing a pre-run sounds good)
- hooks
- parametrization
- checkpointing

> Finally, I am wondering how to deal with DP/DDP modules.

Once we have a functional nn.Module, it is going to be easy to re-implement these wrappers.
We will need to have a special "replication" function for these Modules that will most likely replicate the wrapped Module and then wrap that into a simple function that run these in parallel.

> believe that changes to torch.nn.Module might be needed in order to do a clean implementation and just don't do another monkey patching approach.

I think this is fine to do.
I think that the workflow of Proof of Concept then refactor then implement will work fine here. As we will need a full prototype to make sure it works fine with all the other components and that might require some cleanup.

**Comment by zou3519** [comment]
I don't think I posted this in this thread already, but there's [a prototype of make_functional in functorch](https://github.com/facebookresearch/functorch/blob/946ff209405e860d93662872a26e834d9ca1c36e/functorch/_src/make_functional.py#L238-L330) that we could use as a starting point and run with

**Comment by ezyang** [comment]
cc @jbschlosser
